feat(Nag Suppressions): granular suppressions

tkglaser opened the pull request:
Related to [#634](https://github.com/cdklabs/cdk-nag/issues/634)
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

tkglaser commented:
Not sure why this failed, the read permission on my fork should be configured and it is marked as public?

dontirun commented:
The GitHub action is running from this repository and does not have permission (rightfully so) to push changes to the fork. In this case the `API.md` generated from the build process differed from what was pushed to the fork so it attempted to fix that by pushing the new documentation.

To fix that, run the `npx projen build`  command and push the new `API.md` to your fork.

tkglaser commented:
You're right, I forgot to run the build, sorry. I've pushed an update.

dontirun commented on RULES.md:
```diff
@@ -92,7 +92,7 @@ The [AWS Solutions Library](https://aws.amazon.com/solutions/) offers a collecti
 | AwsSolutions-EVB1  | The event bus policy allows for open access.                                                                                                                                              | An open policy ("\*" principal without a condition) grants anonymous access to an event bus. Use a condition to limit the permission to accounts that fulfill a certain requirement, such as ...
 | AwsSolutions-IAM4  | The IAM user, role, or group uses AWS managed policies.                                                                                                                                   | An AWS managed policy is a standalone policy that is created and administered by AWS. Currently, many AWS managed policies do not restrict resource scope. Replace AWS managed policies with ...
-| AwsSolutions-IAM5  | The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.                                                             | Metadata explaining the evidence (e.g. via supporting links) for wildcard permissions allows for transparency to operators.                                                                  ...
+| AwsSolutions-IAM5  | The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.                                                             | Metadata explaining the evidence (e.g. via supporting links) for wildcard permissions allows for transparency to operators. This is a granular rule that returns indivisual findings that ca ...
```
This should also be reflected [here](https://github.com/tkglaser/cdk-nag/blob/d0901a521e2c042075e3d0666589de9cc2659b90/src/packs/aws-solutions.ts#L1209)

dontirun reviewed (changes requested):
I think we should throw an error if a users includes a finding in the suppression `id` and point them to documentation on how to properly use findings.

We could add something like this in the `NagSuppressionHelper`
``` ts
static validateSuppressions(id: string, suppressions: NagPackSuppression[]): NagPackSuppression[] {
  const validatedSuppressions: NagPackSuppression[] = []
  const errors = []
  for (const suppression of suppressions) {
    const errorMessage = NagSuppressionHelper.getSuppressionFormatError(suppression)
    errorMessage ? errors.push(errorMessage) : validatedSuppressions.push(suppression);
  }
  if (errors.length) {
    throw Error(
      `${id}: ${errors.join('')}\nSee https://github.com/cdklabs/cdk-nag#suppressing-a-rule for information on suppressing a rule.`
    );
  }
  return validatedSuppressions
}
static getSuppressionFormatError(suppression: NagPackSuppression): string {
  let errors = ''
  const finding = suppression.id.match(/\[.*\]/)
  if (finding) {
    errors += `The suppression 'id' contains a finding '${finding}. A finding must be suppressed using 'applies_to'.`
  }
  if (suppression.reason.length < 10) {
    errors += "The suppression must have a 'reason' of 10 characters or more."
  }
  return errors ? `\n\tError(s) detected in suppression with 'id' ${suppression.id}. ${errors}` : ''
}
```

and update [addStackSuppressions](https://github.com/tkglaser/cdk-nag/blob/d0901a521e2c042075e3d0666589de9cc2659b90/src/nag-suppressions.ts#L29) to
``` ts
const newSuppressions = NagSuppressionHelper.validateSuppressions(s.node.id, suppressions);
```

 and update [addResourceSuppressions](https://github.com/tkglaser/cdk-nag/blob/d0901a521e2c042075e3d0666589de9cc2659b90/src/nag-suppressions.ts#L63) to 
 
 ```ts
const newSuppressions = NagSuppressionHelper.validateSuppressions(construct.node.id, suppressions);

 ```
 
 We can then add an example on suppressing findings in the `README` 
 
 That way a suppression like this

```ts
NagSuppressions.addResourceSuppressions(
  user,
  [
    {
      id: 'AwsSolutions-IAM5[Resource::*]',
      reason: 'Every finding is suppressed.',
    },
    {
      id: 'AwsSolutions-IAM5[Action::as]',
      reason: 'Every finding is suppressed.',
    },
  ],
  true
    );
```

will throw an Error like this
```bash
  rUser: 
      Error(s) detected in suppression with 'id' AwsSolutions-IAM5[Resource::*]. The suppression 'id' contains a finding '[Resource::*]. A finding must be suppressed using 'applies_to'.
      Error(s) detected in suppression with 'id' AwsSolutions-IAM5[Action::as]. The suppression 'id' contains a finding '[Action::as]. A finding must be suppressed using 'applies_to'.
  See https://github.com/cdklabs/cdk-nag#suppressing-a-rule for information on suppressing a rule.
```

tkglaser commented:
Good idea! 👍

dontirun reviewed (changes requested):
Let's also update rule [compliance validation](https://github.com/tkglaser/cdk-nag/blob/208295cf7ae6a01f909c8338bc19c4198a66fef6/test/rules/utils.ts#L21) for testing rules to check for the findings format.

```ts 
data: expect.stringMatching(`.*${ruleId}(\\[.*\\])?:`),
```

tkglaser commented:
Update pushed:
* I've simplified the rule `validateSuppressions` method a little. It's not actually necessary to return an array, as the array is either identical to the input or the method throws an error
* An example on suppressing findings was already in the readme but I can add more if you want
* Added a CFN example to the readme
* The change to utils would have prevented us from checking for specific findings in unit tests so that would have broken some of the IAM5 unit tests that I had changed as part of this PR

dontirun commented:
I think we should have an explicit callout in the README [here](https://github.com/tkglaser/cdk-nag/tree/feature/granular-suppressions#suppressing-a-rule).  in addition to the examples you've added.

Somthing like the following

<details>
  <summary>Example 5) Granular Suppressions of findings</summary>

Certain rules support granular suppressions of `findings`. If you received the following errors on synth/deploy

```bash
[Error at /StackName/rFirstUser/DefaultPolicy/Resource] AwsSolutions-IAM5[Action::s3:*]: The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.
[Error at /StackName/rFirstUser/DefaultPolicy/Resource] AwsSolutions-IAM5[Resource::*]: The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.
[Error at /StackName/rSecondUser/DefaultPolicy/Resource] AwsSolutions-IAM5[Action::s3:*]: The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.
[Error at /StackName/rSecondUser/DefaultPolicy/Resource] AwsSolutions-IAM5[Resource::*]: The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.
```

By applying the following suppressions

```typescript
import { User } from '@aws-cdk/aws-iam';
import { NagSuppressions } from 'cdk-nag';
import { Construct, Stack, StackProps } from '@aws-cdk/core';

export class CdkTestStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);
    const firstUser = new User(this, 'rFirstUser');
    firstUser.addToPolicy(
      new PolicyStatement({
        actions: ['s3:*'],
        resources: ['*'],
      })
    );
    const secondUser = new User(this, 'rSecondUser');
    secondUser.addToPolicy(
      new PolicyStatement({
        actions: ['s3:*'],
        resources: ['*'],
      })
    );
    NagSuppressions.addResourceSuppressions(
      firstUser,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: "Only suppress AwsSolutions-IAM5 's3:*' finding on First User.",
          appliesTo: ['Action::s3:*'],
        },
      ],
      true
    );
    NagSuppressions.addResourceSuppressions(
      secondUser,
      [
        {
          id: 'AwsSolutions-IAM5',
          reason: "Suppress all AwsSolutions-IAM5 findings on Second User.",
        },
      ],
      true
    );
  }
}
```

You would see the following error on synth/deploy

```bash
[Error at /StackName/rFirstUser/DefaultPolicy/Resource] AwsSolutions-IAM5[Resource::*]: The IAM entity contains wildcard permissions and does not have a cdk_nag rule suppression with evidence for those permission.
```

</details>

dontirun commented:
>     * The change to utils would have prevented us from checking for specific findings in unit tests so that would have broken some of the IAM5 unit tests that I had changed as part of this PR

The changes to utils would only affect the `TestType.COMPLIANCE` tests, not the `TestType.NON_COMPLIANCE` tests. Without that change we might have some rules with findings that accidentally pass compliance tests because of how the `[finding]` changes the message output.

tkglaser commented:
Update pushed:
* Added the example to the readme
* Made the change to the utils. You were right, when I tried it I made the change to both `COMPLIANCE` and `NON_COMPLIANCE` which broke it. I've now only changed `COMPLIANCE` which works.

dontirun commented:
Thanks for the contribution! 🥳
